Add July 2021 minutes (#144)
tc39/tc39-notes: meetings/2021-07/july-14.md

# 14 July, 2021 Meeting Notes
-----

**Remote attendees:** 
| Name                 | Abbreviation   | Organization       |
| -------------------- | -------------- | ------------------ |
| Waldemar Horwat      | WH             | Google             |
| Istvan Sebestyen     | IS             | Ecma International |
| Richard Gibson       | RGN            | OpenJS Foundation  |
| Jack Works           | JWK            | Sujitech           |
| Josh Blaney          | JPB            | Apple              |
| Chengzhong Wu        | CZW            | Alibaba            |
| Devin Rousso         | DRO            | Apple              |
| Philip Chimento      | PFC            | Igalia             |
| Jamie Kyle           | JK             | Rome               |

## Ergonomic Brand Checks for Stage 4
Presenter: Jordan Harband (JHD)

- [proposal](https://github.com/tc39/proposal-private-fields-in-in/)
- [slides](https://github.com/tc39/proposal-private-fields-in-in/issues/7)

> JHD: So, as a quick refresher, class fields is a stage 4 feature. However, if you would like to determine if an object has a private field, currently you have to use try-catch. This proposal gives you a straightforward boolean check of whether a private field is in object and that includes fields, private methods, and private accessors. The proposal stage 3, and I have all of these check marks here to measure progress toward stage four. So, as of this morning, Firefox version 90 is available Chrome. 91 has been available for a while. With this feature. The features also in webkit nightly and should be in the next Safari technology preview, which I believe will make it into the next full version of safari. But of course, Apple never discloses that before it's released. It's in node as a version 16.4.0. And it's also will be default. It's been in Babel for a while, it will be on by default with the end preset as soon as someone releases the package and merges the pr, which will happen after this feature gets stage the pull request of the spec exists and is approved. Test 262 test also have been landed. And so based on all of that I believe it clearly meets the criteria for stage four and want to ask the committee for consensus for that.

Reviewer: ```suggestion
JHD: So, as a quick refresher, class fields is a stage 4 feature. However, if you would like to determine if an object has a private field, currently you have to use try-catch. This proposal gives you a straightforward boolean check of whether a private field is in an object and that includes fields, private methods, and private accessors. The proposal is stage 3, and I have all of these checkmarks here to measure progress toward stage 4. So, as of this morning, Firefox version 90 is available, and Chrome 91 has been available for a while, with this feature. The feature is also in Webkit Nightly and should be in the next Safari Technology Preview, which I believe will make it into the next full version of Safari. (But of course, Apple never discloses that before it's released). It's in node as of v16.4.0. It's been in Babel for a while; it will be on by default with the env preset as soon as someone releases the package and merges the pr, which will happen after this feature gets stage 4. The pull request to the spec exists and is approved. test262 tests also have been landed. And so based on all of that I believe it clearly meets the criteria for stage 4 and want to ask the committee for consensus for that.
```

DDC: I mean that would be another alternative that I think could allow, like, it could be say allow asserts type equals JavaScript's. That that's basically the same thing as not having the type of solution and then the rest of the set of reserved strings is banned and causes a failure. I think that would also be reasonable. We'd also I think at some point has been discussed that like, like we discussed whether or not a "type JavaScript" assertion should be supported and I think we'd elected not to because it's better not have two ways to do the same thing, but I think it would be reasonable to allow one such assertion. It's just in any case not - I think that's almost a separate question like whether something like, like, this should be allowed is like one question of whether that's a type of solution that should work in say the web and then there's kind of a separate question of whether we should have a block list of all the types that aren't that, or like all the types including that.

SYG: Yeah, I think I agree that these are orthogonal concerns.

> JHD: I don't think there's any way we can phrase this that won't be relatively trivial to bypass. If we have a list, then someone will come up with a new way to mean JavaScript, right? Like JS, JavaScript, jscript. I don't know. Someone will come up with something, but then like, if it's if we do the like prose option, I mean Source, Testament, text module records, I can make my own kind of module record as a host that is a source text module record but with some optional trivial difference, which means effectively, includes all sorts text module records and then I can bypass this assertion entirely and still be in compliance. So, like, I think that on some level, I don't think any host is likely to do this, and if they do they wouldn't be able to leverage almost all the code out there that's using no assertions to bring in JavaScript. So like it kind of seems - it seems very unlikely to me that without a restriction this would become an interop concern because I can't see enough people using such a host with that assertion. So it kind of seems like the way things are is ideal.

Reviewer: ```suggestion
JHD: I don't think there's any way we can phrase this that won't be relatively trivial to bypass. If we have a list, then someone will come up with a new way to mean JavaScript, right? Like JS, JavaScript, jscript. I don't know. Someone will come up with something, but then like, if it's if we do the like prose option, I mean Source Text Module Records, I can make my own kind of module record as a host that is a source text module record but with some optional trivial difference, which means effectively, includes all sorts text module records and then I can bypass this assertion entirely and still be in compliance. So, like, I think that on some level, I don't think any host is likely to do this, and if they do they wouldn't be able to leverage almost all the code out there that's using no assertions to bring in JavaScript. So like it kind of seems - it seems very unlikely to me that without a restriction this would become an interop concern because I can't see enough people using such a host with that assertion. So it kind of seems like the way things are is ideal.
```

DDC: I think I followed you until the very end. The first part of that I kind of agree that a determined host could get around this if they wanted to. But then like you were saying in practice that they might probably not want to do that because that's going to be like, why? Because you're going to have those sorts of interop issues. Could you restate the conclusion?

JHD: So My personal intuition here, and obviously I could be wrong, is that if a host chose to make this JavaScript assertion required they would quickly find themselves unused by the population. And if a host decided to make it optional then authors of those modules who used it to import their transitive dependencies would quickly find their modules going unused by the larger population. My suspicion is that anyone attempting to allow this sort of thing is not going to gain enough popularity to then be an interop concern.

BT: All right, the queue is empty.

### Conclusion/Resolution
No changes sought or made, general agreement on the status quo

## Decorators update
Presenter: Chris Garrett (CHG)

> - [proposal](https://github.com/tc39/proposal-decorators

Reviewer 2: missing closing `)`

MM: Okay? So the privates, lot of the instance still exists and if it's still where the private value is stored,

DDC: yes, it does not to change the shape of the class in any way and it doesn't change the way the class works at all. If you don't use the access object or those functions, then nothing changes at all. It's like they were never there. Using them is just a way to access the values in a separate way.

MM: And if the getter and setter is applied to anything other than an instance, it throws?

DDC: it should. If that's not in the spec then I should definitely add it. Any other questions?

JHD: Yeah, it was a couple slides back where you're showing the decorator function interface. Yeah, here you go. I was just curious if the context access set function, can that ever fail, or it was that only there on a private field and so like it should never fail.

DDC: If it is called on an instance of the class, because the class should always have that, it should never fail.

> JHD: It should never fail, but if it's called on the instead something else, so it'll either throw because the field isn't on the receiver. The receiver or it will succeed. Is that correct?

Reviewer: ```suggestion
JHD: It should never fail, but if it's called instead on something else, so it'll either throw because the field isn't on the receiver, or it will succeed. Is that correct?
```

DDC: Okay. So that's an overview of the proposal. Any questions about auto accessors? 

BT: There are a few questions on the queue, but also just a quick time note. We're down about eight minutes in the time box, I believe. So try to be brief.

JHD: So two questions. The first was with the auto accessor feature that you described here, would this a viable standalone proposal, that could move independently as long as it was faster than independently of decorators and have value on its own, or do you think it only has value with decorators?

DDC: I think it could move independently on its own. I definitely think that there are a lot of existing decorator use cases that depend on the ability to intercept access on fields. So either this proposal would have to kind of really be concurrent or ideally land first, if it were to be a separate proposal.

> JHD :Okay. And then the second question was with your "similar to". I was curious how faithful that sugar was. In other words, will the getter throw if it's called an object that isn't it? Like the didn't run through the constructor? You know like where as if it was a private field like your line 6 through 16 here will throw if the receiver lacks that private field. So I'm wondering if the auto accessors works the same. Then if so, how would I determine in advance without a try/catch if the accessor is going to throw or not?

Reviewer: ```suggestion
JHD: Okay. And then the second question was with your "similar to". I was curious how faithful that sugar was. In other words, will the getter throw if it's called an object that isn't it? Like the didn't run through the constructor? You know like where as if it was a private field like your line 6 through 16 here will throw if the receiver lacks that private field. So I'm wondering if the auto accessors works the same. Then if so, how would I determine in advance without a try/catch if the accessor is going to throw or not?
```

DDC: That is a great question. The second one, rather. The first one is the answer is it will throw, I believe, because in the spec we back the field with a private slot, it uses all the same mechanisms as a private field. For the second question, I'm not too sure. I think you would have to use a try/catch. I'd have to think on that or what it means that like General concern, 

> JHD: there is a like address the proposal I had a few ago that just it stage 4 is like I wanted to determine if something is going to throw or not with the in keyword so that works for private Fields. But in this case because it's sugar for it, there isn't Special way to refer to it that would work within and tell you if it's going to throw or not. Because if did some value in an object, it'll just tell you if the accessors, you know, if the property is successful there, right?

Reviewer: ```suggestion
JHD: there is the proposal I discussed recently that just got stage 4 is that I wanted to determine if something is going to throw or not with the `in` keyword, so that works for private fields now. But in this case because it's sugar for it, there isn't a special way to refer to it that would work within and tell you if it's going to throw or not. Because if did some value in an object, it'll just tell you if the accessors, you know, if the property is successful there, right?
```

PFC: I can share it right now: https://github.com/tc39/proposal-temporal/issues/1583

SYG: I support the proposal as-is, as in the slide that you had open, which is fully qualified name in the @@toStringTag and then the capitalization. I may be wrong because I haven't done research here across a set of other languages, but the built-in names for things, my intuition across most programming languages, is that the conventions for how they are capitalized or spelled usually is derived from the kind of thing they are, or how they're meant to be used, and not really where they are put, like if they're nested or not. For that reason, I support that namespace objects, like constructors, remain capitalized, regardless of where they get installed by default.

JRL: `Date.now()` is a method that you call and having `Temporal.now` be lowercase makes me think it's supposed to be a callable function that's going to give me like date, time, or something in the Temporal object. Making it a capital namespace is going to prevent me from trying to just call it randomly.

> JHD: This is is a counter to my own constructor comment as well, but if you think it's a thing that it's not, and you try to use it that way, and then it immediately tells you it's not a function, you can't do something. If it's not a function, and you can't call it, you're not going to be confused anymore. So I don't know if the confusion argument really pushes in either direction very strongly.

Reviewer: ```suggestion
JHD: This is a counter to my own constructor comment as well, but if you think it's a thing that it's not, and you try to use it that way, and then it immediately tells you it's not a function, you can't do something. If it's not a function, and you can't call it, you're not going to be confused anymore. So I don't know if the confusion argument really pushes in either direction very strongly.
```

CM: I think while the proposal is well-intentioned, I think it kind of misses the point of the original terminology, the idea of a straw man or straw person was something which is kind of initially flimsy, which is what things often are at the beginning when they haven't been fleshed out. And the whole idea is something which is in a flimsy or weak state. And then as you work on it, you make it stronger and more sturdy and more sound. A “concept” is just an idea. The whole idea behind "straw person" or whatever word you use for it, is the idea that you're capturing there, is that it is quite possibly very flawed: it's very flimsy, it's not necessarily all the way thought through and should be regarded such, meaning it'll be easy to attack and therefore the thing to do is not just blindly attack it, because anybody can do that, but to have constructive contributions that can shore it up and make it stronger, make it more sound. And that sort of evolution from something, which starts as a sort of weak proto-notion to something which is more fleshed out is, I think, lost if you switch to some abstraction like "concept".

AKI: Thanks for that chip. That's a good point.

> SYG: So maybe I misunde the actual intention of this proposal. Is the proposal to change the process document or to kind of get us on the same page to try to stop using the word straw person?

Reviewer 3: ```suggestion
SYG: So maybe I misunderstand the actual intention of this proposal. Is the proposal to change the process document or to kind of get us on the same page to try to stop using the word straw person?
```

AKI: I think it's both.

> SYG: I see. It seems like it would be clear for the process document to perhaps just drop any mnemonic.and just, we just refer to Stage 0 through 4. As for getting us on the same page, that seems like a taller order but because I have nothing against "concept". Sorry, that was not very clear. I'm proposing that instead of spending time in plenary to debate whether "concept" is a good replacement for strawperson, we could just drop it from the process document.

Reviewer 3: ```suggestion
SYG: I see. It seems like it would be clear for the process document to perhaps just drop any mnemonic just refer to Stage 0 through 4. As for getting us on the same page, that seems like a taller order but because I have nothing against "concept". Sorry, that was not very clear. I'm proposing that instead of spending time in plenary to debate whether "concept" is a good replacement for strawperson, we could just drop it from the process document.
```

HHM: So the general observation here, as most of them are comfortable, calling it by State 0 to stage 4, or rather than names, and the contrasting opinion is we have names, then it would be easier for people to explain it to newcomers, or folks who don't have idea about stages. The other opinion is like, if we call it stage zero, then they will probably go and look look into what "stage 0" means. So, I think the majority of folks are comfortable calling it stage 0 through stage 4 rather than by names. So should we get rid of names?

SFC: I have two queue entries. First is, a Stage 0 proposal is not really formalized until it reaches Stage 1; it's not really even a proposal, it's almost more like a suggestion, because it's not really a proposal until it actually reaches Stage 1. So like if we were to keep the English names, I feel like Stage 1 being called "concept", Stage 0 being called something like "suggestion" or something like that.

> JHD: It's a proposal when a TC39 delegate champions it and plans to present it, that's usually the bar we use and that has been useful for things that did not make it to stage one, because they have come back again and they're still useful as reference points. So I think it's still useful to list stage 0.

Reviewer: ```suggestion
JHD: It's a proposal when a TC39 delegate champions it and plans to present it, that's usually the bar we use and that has been useful for things that did not make it to stage 1, because they have come back again and they're still useful as reference points. So I think it's still useful to list stage 0.
```

AKI: Can I just repeat so I understand properly and in case other people are also not understanding? What you're saying is keeping stage zero on the process document is valuable because even though we're not committed to it, we can track that somebody did the work even if it never made it into the proper staging process. Is that what you're saying Jordan?

> JHD: Yeah. I mean I'm saying that the having a stage zero, having a perp repo for a proposal that is still in stage 0 but that has has or will get committee times as useful as a place to concentrate attention and track things, sure.

Reviewer: ```suggestion
JHD: Yeah. I mean I'm saying that having a repo for a proposal that is still in stage 0 but that has or will get committee time is useful as a place to concentrate attention and track things, sure.
```

MF:  I'm fine with that. Please Justin in your description in your repository, address the problem and not just the solution.

WH: I take a different procedural position and I would say that, in my opinion, the `groupBy` proposal is almost at stage 2. We already have spec text for it, with the only modulo being that I would want the prototype gone from the produced objects and possibly a Map version. 

JRL: So, I think with agreement that the Prototype object has to be changed to use object.create(null), or we have to use a Map, or a solution that does both. I think that's an agreement. There's also holes, which I think we still need to discuss but we can do that at stage one, and naming, which I think we're in agreement that just normal "groupBy" is fine. I think that is what I'm proposing to go to the stage one, that agreement

> JHD: I completely agree with the point making but I also think if we're going to consider both guilt or reject and groupBy separately, then they need to have two separate repositories for like Clarity and organization and whe named those proposals and what stage we decide there is a separate thing. But like, I think it's useful to have a specific repo that talks about group. I The problems at solving for the dresses, the Prototype thing and so on. And and I. So I think because of that, we need, it's prudent for us to separate it. Whether that's goes at stage one or two.

Reviewer: ```suggestion
JHD: I completely agree with the point making but I also think if we're going to consider both filterReject and groupBy separately, then they need to have two separate repositories for clarity and organization, and what we name those proposals and what stage we decide they are is a separate thing. But I think it's useful to have a specific repo that talks about groupBy, the problems it is solving or addresses, the prototype thing, and so on. So I think because of that, we need, it's prudent for us to separate it. Whether that goes at stage 1 or 2.
```
